packages/cosmic-swingset/src/helpers/input-worker.js

```diff
@@ -0,0 +1,124 @@
+// @ts-check
+import { Far } from '@endo/far';
+import { makeReadlineLoop } from './read-fd-lines.js';
+import { makeJsonRPCInterceptor } from './interceptor.js';
+import stableStringify from './json-stable-stringify.js';
+import {
+  DEFAULT_BUFFER_SIZE,
+  makeTextEncoder,
+  makeTextDecoder,
+  makeTransferClient,
+  makeTransferServer,
+} from './sync-xfer.js';
+
+const debug = (..._args) => {}; // console.error(..._args);
+
+/**
+ * @param {import('worker_threads').MessagePort} parent
+ */
+export const makeInputWorkerHandler = parent => {
+  /** @type {ReturnType<makeJsonRPCInterceptor>} */
+  let intercept;
+  return data => {
+    debug('message handler received', data);
+    switch (data && data.type) {
+      case 'INIT_LOOP': {
+        const { fd } = data;
+        if (intercept) {
+          throw Error('Already initialized');
+        }
+        intercept = makeJsonRPCInterceptor(obj =>
+          parent.postMessage({ type: 'RECV_OBJ', fd, obj }),
+        );
+        makeReadlineLoop(fd, intercept.handler)
+          .then(intercept.resolve, intercept.reject)
+          .finally(() => parent.postMessage({ type: 'FINISH_LOOP', fd }));
+        break;
+      }
+
+      case 'INTERCEPT_ID': {
+        /** @type {{ id: number, buf: Int32Array }} */
+        const { id, buf } = data;
+        const xferServer = makeTransferServer(buf);
+        const interceptor = obj => {
+          debug('intercepted id', id, obj);
+          xferServer.write(makeTextEncoder(stableStringify(obj)));
```

Comment:
Depending upon GC sophistication, it might be necessary here to explicitly drop references.
```suggestion
          xferServer.write(makeTextEncoder(stableStringify(obj)));
          xferServer = interceptor = null;
```